File: cepi-ios/CEPITelemedicina/App/EliminarCuenta.swift
import SwiftUI

/// Borrar la cuenta desde la app (App Store 5.1.1(v)): el login con Google crea cuentas, así
/// que también tienen que poder borrarse acá. Lo usan el menú de la cuenta y la pantalla de
/// cuenta pendiente, que es donde cae una cuenta recién creada con Google.
///
/// El botón vive en quien lo muestra; la confirmación y el error, acá. Van por fuera del
/// botón porque una alerta colgada de un botón de `Menu` no llega a presentarse: el menú se
/// cierra y se lleva la vista.
struct EliminarCuenta: ViewModifier {
    @Environment(Sesion.self) private var sesion
    @Binding var confirmar: Bool
    @State private var eliminando = false
    @State private var error: String?

    static let explicacion = """
        Se borran tu email, teléfono y cédula, tu acceso con contraseña o con Google, tus \
        organizaciones y las notificaciones en tus dispositivos. No se puede deshacer.

        Las historias clínicas que registraste no se borran: pertenecen al paciente y a la \
        institución, y la ley obliga a conservarlas. Por eso en ellas sigue tu nombre como \
        profesional que atendió.
        """

    func body(content: Content) -> some View {
        content
            .disabled(eliminando)
            .overlay {
                if eliminando { ProgressView() }
            }
            .alert("¿Eliminar tu cuenta?", isPresented: $confirmar) {
                Button("Eliminar cuenta", role: .destructive) {
                    Task { await eliminar() }
                }
                Button("Cancelar", role: .cancel) {}
            } message: {
                Text(Self.explicacion)
            }
            .alert("No se pudo eliminar la cuenta", isPresented: Binding(
                get: { error != nil },
                set: { if !$0 { error = nil } }
            )) {
                Button("Aceptar", role: .cancel) {}
            } message: {
                Text(error ?? "")
            }
    }

    private func eliminar() async {
        eliminando = true
        defer { eliminando = false }
        do {
            try await sesion.eliminarCuenta()
        } catch {
            self.error = error.localizedDescription
        }
    }
}

extension View {
    /// Confirmación y error del borrado de cuenta; `confirmar` lo pone en `true` el botón.
    func eliminarCuenta(confirmar: Binding<Bool>) -> some View {
        modifier(EliminarCuenta(confirmar: confirmar))
    }
}

File: cepi-android/app/src/main/java/ec/cepi/telemedicina/app/EliminarCuenta.kt
package ec.cepi.telemedicina.app

import androidx.compose.foundation.layout.size
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import ec.cepi.telemedicina.api.ApiError
import kotlinx.coroutines.launch

private const val EXPLICACION =
    "Se borran tu email, teléfono y cédula, tu acceso con contraseña o con Google, tus " +
        "organizaciones y las notificaciones en tus dispositivos. No se puede deshacer.\n\n" +
        "Las historias clínicas que registraste no se borran: pertenecen al paciente y a la " +
        "institución, y la ley obliga a conservarlas. Por eso en ellas sigue tu nombre como " +
        "profesional que atendió."

/**
 * Borrar la cuenta desde la app (Play y App Store lo exigen): el login con Google crea
 * cuentas, así que también tienen que poder borrarse acá. Lo usan el menú de la cuenta y la
 * pantalla de cuenta pendiente. Si el servidor no borra, la sesión sigue y el motivo se muestra
 * tal cual.
 */
@Composable
fun DialogoEliminarCuenta(sesion: Sesion, alCerrar: () -> Unit) {
    val alcance = rememberCoroutineScope()
    var eliminando by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }

    val fallo = error
    if (fallo != null) {
        AlertDialog(
            onDismissRequest = alCerrar,
            title = { Text("No se pudo eliminar la cuenta") },
            text = { Text(fallo) },
            confirmButton = { TextButton(onClick = alCerrar) { Text("Aceptar") } },
        )
        return
    }

    AlertDialog(
        onDismissRequest = { if (!eliminando) alCerrar() },
        title = { Text("¿Eliminar tu cuenta?") },
        text = { Text(EXPLICACION) },
        confirmButton = {
            TextButton(
                enabled = !eliminando,
                onClick = {
                    eliminando = true
                    alcance.launch {
                        try {
                            sesion.eliminarCuenta()
                        } catch (e: ApiError) {
                            error = e.mensaje
                        } finally {
                            eliminando = false
                        }
                    }
                },
            ) {
                if (eliminando) {
                    CircularProgressIndicator(Modifier.size(18.dp), strokeWidth = 2.dp)
                } else {
                    Text("Eliminar cuenta", color = MaterialTheme.colorScheme.error)
                }
            }
        },
        dismissButton = {
            TextButton(onClick = alCerrar, enabled = !eliminando) { Text("Cancelar") }
        },
    )
}
